### PR TITLE
Move user blurring logic into UserDimContainer

### PR DESCRIPTION
**Pull request by nyquillerium**
This moves all user-specified blur behavior into UserDimContainer, now renamed to VisualSettingsContainer. This makes it so that all user-defined settings logic are in one place.

Also adds some asserts to TestCaseBackgroundScreenBeatmap to test this new behavior.

 - Closes https://github.com/ppy/osu/issues/4322

### CODE REVIEW

**Comment by nyquillerium** [comment]
As discussed irl, going to rework this so it doesn't require nesting buffered containers for performance reasons.

File: osu.Game/Screens/Play/ScreenWithBeatmapBackground.cs [hunk 1]
```diff
@@ -17,49 +14,12 @@ public abstract class ScreenWithBeatmapBackground : OsuScreen
...
-        protected Bindable<double> BlurLevel;
         protected Bindable<bool> ShowStoryboard;
 
-        #endregion
-
         [BackgroundDependencyLoader]
         private void load(OsuConfigManager config)
         {
-            BlurLevel = config.GetBindable<double>(OsuSetting.BlurLevel);
             ShowStoryboard = config.GetBindable<bool>(OsuSetting.ShowStoryboard);
```
Comment (peppy) [inline osu.Game/Screens/Play/ScreenWithBeatmapBackground.cs]: What is this doing?

**Comment by peppy** [comment]
Did we decide on thename `VisualSettingsContainer`? Personally I think this doesn't fit well as there is also `VisualSettings`, which will likely contain more settings than just the ones managed by the said container.

@smoogipoo thoughts? I think `UserDimContainer` works better and still consider "blur" to be part of the user dim function.

File: osu.Game/Graphics/Containers/UserDimContainer.cs [hunk 2]
```diff
@@ -38,13 +42,19 @@ public class UserDimContainer : Container
 
         private readonly bool isStoryboard;
 
+        public Bindable<float> AddedBlur = new Bindable<float>();
```
Comment (peppy) [inline osu.Game/Graphics/Containers/UserDimContainer.cs]: xmldoc

Comment (peppy) [reply osu.Game/Graphics/Containers/UserDimContainer.cs]: also move all the private bindables below public ones

File: osu.Game/Graphics/Containers/UserDimContainer.cs [hunk 3]
```diff
@@ -38,13 +42,19 @@ public class UserDimContainer : Container
...
+
+        public Vector2 BlurTarget => EnableVisualSettings.Value
```
Comment (peppy) [inline osu.Game/Graphics/Containers/UserDimContainer.cs]: does this need to be more than private?

File: osu.Game/Screens/Play/Player.cs [hunk 4]
```diff
@@ -73,11 +73,13 @@ public class Player : ScreenWithBeatmapBackground
         private DrawableStoryboard storyboard;
         protected UserDimContainer StoryboardContainer { get; private set; }
 
+        private Bindable<bool> showStoryboard;
+
         protected virtual UserDimContainer CreateStoryboardContainer() => new UserDimContainer(true)
         {
             RelativeSizeAxes = Axes.Both,
             Alpha = 1,
-            EnableUserDim = { Value = true }
```
Comment (peppy) [inline osu.Game/Screens/Play/Player.cs]: can we rename this back too?

**Review by peppy — CHANGES_REQUESTED** [review]

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/BackgroundScreen.cs [hunk 5]
```diff
@@ -5,12 +5,15 @@
...
 using osu.Framework.Graphics;
 using osu.Framework.Input.Events;
+using osu.Game.Graphics.Backgrounds;
 using osuTK;
 
 namespace osu.Game.Screens
 {
     public abstract class BackgroundScreen : Screen, IEquatable<BackgroundScreen>
     {
+        protected Background Background;
```
Comment (peppy) [inline osu.Game/Screens/BackgroundScreen.cs]: This can't just exist here like this. Properties in a class should be used by the class itself. Either `BackgroundScreen` should be responsible for setting this or it should be added in a per-class basis. Otherwise there's little point to this existing.

File: osu.Game/Screens/Backgrounds/BackgroundScreenBeatmap.cs [hunk 6]
```diff
@@ -11,17 +11,19 @@
...
-    public class BackgroundScreenBeatmap : BlurrableBackgroundScreen
+    public class BackgroundScreenBeatmap : BackgroundScreen
     {
         private WorkingBeatmap beatmap;
 
         /// <summary>
         /// Whether or not user dim settings should be applied to this Background.
         /// </summary>
-        public readonly Bindable<bool> EnableUserDim = new Bindable<bool>();
+        public readonly Bindable<bool> EnableVisualSettings = new Bindable<bool>();
```
Comment (peppy) [inline osu.Game/Screens/Backgrounds/BackgroundScreenBeatmap.cs]: rename back

File: osu.Game/Screens/Backgrounds/BackgroundScreenBeatmap.cs [hunk 7]
```diff
@@ -11,17 +11,19 @@
...
 
         public readonly Bindable<bool> StoryboardReplacesBackground = new Bindable<bool>();
 
+        public readonly Bindable<float> AddedBlur = new Bindable<float>();
```
Comment (peppy) [inline osu.Game/Screens/Backgrounds/BackgroundScreenBeatmap.cs]: xmldoc

Comment (peppy) [reply osu.Game/Screens/Backgrounds/BackgroundScreenBeatmap.cs]: also rename to just `BlurAmount` or something.

File: osu.Game/Screens/Play/PlayerLoader.cs [hunk 8]
```diff
@@ -118,12 +119,18 @@ private void loadNewPlayer()
...
+            Background.AddedBlur.Value = BACKGROUND_BLUR;
+            Background.EnableVisualSettings.Value = false;
+
             content.ScaleTo(1, 650, Easing.OutQuint);
             content.FadeInFromZero(400);
         }
 
         private void contentOut()
         {
+            Background.AddedBlur.Value = 0;
```
Comment (peppy) [inline osu.Game/Screens/Play/PlayerLoader.cs]: this code exists in two places

File: osu.Game/Screens/Play/PlayerLoader.cs [hunk 9]
```diff
@@ -118,12 +119,18 @@ private void loadNewPlayer()
 
         private void contentIn()
         {
+            Background.AddedBlur.Value = BACKGROUND_BLUR;
```
Comment (peppy) [inline osu.Game/Screens/Play/PlayerLoader.cs]: this code exists in two places

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Play/PlayerLoader.cs [hunk 10]
```diff
@@ -172,22 +179,16 @@ protected override void OnHoverLost(HoverLostEvent e)
         {
             if (GetContainingInputManager()?.HoveredDrawables.Contains(VisualSettings) == true)
             {
-                // Update background elements is only being called here because blur logic still exists in Player.
```
Comment (peppy) [inline osu.Game/Screens/Play/PlayerLoader.cs]: whyyyyyyyyyy is this gone??????
it's a very important comment.

Comment (nyquillerium) [reply osu.Game/Screens/Play/PlayerLoader.cs]: update background elements isn't being called anymore though, so its no longer correct.
Also the line afterwards explicitly mentions the issue this resolves so... not sure what you mean

Comment (peppy) [reply osu.Game/Screens/Play/PlayerLoader.cs]: a comment should be added back to explain why the onhover/hoverlost is applied to the screen and not the visual settings overlay.

Comment (nyquillerium) [reply osu.Game/Screens/Play/PlayerLoader.cs]: Added new comments regarding why this hover logic exists

File: osu.Game/Graphics/Containers/UserDimContainer.cs [hunk 11]
```diff
@@ -72,15 +70,13 @@ public UserDimContainer(bool isStoryboard = false)
...
-        /// <remarks>
-        /// We need to support instant blurring here in the case of changing beatmap backgrounds, where blurring shouldn't be from 0 every time the beatmap is changed.
-        /// </remarks>
-        public void ApplyInstantBlur()
+        public override void Add(Drawable drawable)
         {
-            background?.BlurTo(blurTarget, 0, Easing.OutQuint);
+            // We need to blur instantly here in the case of changing beatmap backgrounds, where blurring shouldn't be from 0 every time the beatmap is changed.
+            if (drawable is Background b)
+                b.BlurTo(blurTarget, 0, Easing.OutQuint);
```
Comment (peppy) [inline osu.Game/Graphics/Containers/UserDimContainer.cs]: store to a field instead of using linq to find the background later.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Graphics/Containers/UserDimContainer.cs (hunk 11 above)
Comment (peppy) [reply osu.Game/Graphics/Containers/UserDimContainer.cs]: comment shouldn't mention outer usages (UserDimContainer should not be aware of beatmap usage)

**Comment by nyquillerium** [comment]
Writing a new test for player -> playerloader, should be done shortly

File: osu.Game/Screens/Play/PlayerLoader.cs [hunk 12]
```diff
@@ -40,8 +41,12 @@ public class PlayerLoader : ScreenWithBeatmapBackground
 
         public override bool DisallowExternalBeatmapRulesetChanges => true;
 
+        public override bool HandlePositionalInput => true;
```
Comment (peppy) [inline osu.Game/Screens/Play/PlayerLoader.cs]: why was this added?

Comment (nyquillerium) [reply osu.Game/Screens/Play/PlayerLoader.cs]: Ah this was from when I was trying to use IsHovered, as that can only be updated if I did this. Will revert.

Comment (nyquillerium) [reply osu.Game/Screens/Play/PlayerLoader.cs]: Actually, just realized we still need this for the PlayerLoader to know whether or not we are hovered over it to continue to the next screen. Since we were using OnHover and OnHoverLost before, this wasn't necessary. Will add a comment to explain this.

**Review by peppy — APPROVED** [review]
